File: 101. Symmetric Tree.java
/*
Problem Link: https://leetcode.com/problems/symmetric-tree/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

import java.util.*;

class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        if (root.left == null && root.right == null) {
            return true;
        }
        if (root.left == null || root.right == null) {
            return false;
        }
        if (root.left.val != root.right.val) {
            return false;
        }

        TreeNode temp = (root.left.right != null) ? root.left.right : null;
        root.left.right = (root.right.right != null) ? root.right.right : null;
        root.right.right = temp;
        return (isSymmetric(root.left) && isSymmetric(root.right));
    }
}

File: Problems/101. Symmetric Tree.py
"""
Problem Link: https://leetcode.com/problems/symmetric-tree/
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if (root == None):
            return True
        if (root.left == None and root.right == None):
            return True
        if (root.left == None or root.right == None):
            return False
        if (root.left.val != root.right.val):
            return False

        temp = root.left.right if root.left.right != None else None
        root.left.right = root.right.right if root.right.right != None else None
        root.right.right = temp
        return self.isSymmetric(root.left) and self.isSymmetric(root.right)
